[nfc] Define Circuit Components (#135)
chipsalliance/firrtl-spec: spec.md

Radix-specified integer literals are only usable when constructing hardware integer literals.
Any use in place of an integer is disallowed.

# Circuit Components

Circuit components are the named parts of a module corresponding to hardware.

There are seven **kinds** of circuit components.

> They are: nodes, wires, registers, output ports, input ports, submodule instances, and memories.

Author: I'll keep my eye on this. It's similar to your other comment about how `Passthrough` "needs to be a module" -- but module is being used in a broad sense to include `module` `extmodule` or `intmodule`. This should be detailed somewhere. But perhaps I'll have it in my notes for a follow-up.

Circuit components can be connected together (see [@sec:connects]).

Each circuit component in a module has a type ([@sec:types]).
It is used to determine the legality of connections.

> 

Author: I'll keep that in mind. Fixed wherever I saw it.

The `regreset`{.firrtl} keyword is used to declare a register with a reset.
The `reg`{.firrtl} keyword is used to declare a register without a reset.

Examples:

```firrtl
wire myclock : Clock
reg myreg : SInt, myclock
```

```firrtl
wire myclock : Clock
reg myreg : SInt, myclock
```
```firrtl
wire myclock : Clock
wire myreset : UInt<1>
wire myinit : SInt
regreset myreg : SInt, myclock, myreset, myinit
```

For both variants of register, the type is given after the colon (`:`{.firrtl}).

Semantically, registers become flip-flops in the design.
The next value is latched on the positive edge of the clock.
The initial value of a register is indeterminate (see [@sec:indeterminate-values]).

## Output Ports and Input Ports

The way a module interacts with the outside world is through its output and input ports.

Example:

```firrtl
input myinput : UInt<1>
output myinput : SInt<8>
```

For both variants of port, the type is given after the colon (`:`{.firrtl}).

The two kinds of ports differ in the rules for how they may be connected (see [@sec:connects]).

## Submodule Instances

A module in FIRRTL may contain submodules.

Example:

```firrtl
inst passthrough of Passthrough
```

> This assumes you have a module named `Passthrough`{.firrtl} declared elsewhere in the current circuit.

Author: Changed to :

This assumes you have a `module`, `extmodule`, or `intmodule` named `Passthrough`{.firrtl} declared elsewhere in the current circuit.

Radix-specified integer literals are only usable when constructing hardware
integer literals.  Any use in place of an integer is disallowed.

# Circuit Components

Circuit components are the named objects which may be contained within a module.
They are: nodes, wires, registers, ports, and submodule instances.

TODO: What about `mem`\s?

## Nodes

Nodes are named expressions in FIRRTL.
They represent a small piece of combinational logic.

Example:

```firrtl
    node mynode = and(in, UInt<4>(1))
```

Nodes are given a name and a definition.
Unlike wires, they can only be assigned to at the point they are defined.

> (TODO: Do all values need to be in scope above the point of definition?)

Author: Do all values need to be in scope above the point of definition in a `node`?

Author: cf the TODO just below in the Wires section:

TODO: Circular definitions? c.f. with nodes.

Author: They do.

Radix-specified integer literals are only usable when constructing hardware
integer literals.  Any use in place of an integer is disallowed.

# Circuit Components

Circuit components are the named objects which may be contained within a module.

> They are: nodes, wires, registers, ports, and submodule instances.

Reviewer: what about "verification" statements (printfs, asserts, etc)

Author: This is a good question. Do you have an opinion on this?

Based on the direction I was leaning, I don't think they are. They:
- do not have a principal type, and they
- may not be connected

TODO: What about `mem`s?

## Nodes

Nodes are named expressions in FIRRTL.
They represent a small piece of combinational logic.

Example:

```firrtl
    node mynode = and(in, UInt<4>(1))
```

Nodes are given a name and a definition.
Unlike wires, they can only be assigned to at the point they are defined.
(TODO: Do all values need to be in scope above the point of definition?)
They cannot be connected (see [@sec:connects]).

## Wires

> A wire is a named combinational circuit component that can be connected to and from using connect statements.

Reviewer: can we use the `wire`{.firrtl} syntax?

Author: The concept is wire, the syntax is `wire`{.firrtl}. I considered this, and in this context, it's better to separate the concept from the syntax. (Otherwise, "port" becomes `input` and `output` and "registers" becomes `reg` and `regreset`).

Radix-specified integer literals are only usable when constructing hardware
integer literals.  Any use in place of an integer is disallowed.

# Circuit Components

Circuit components are the named parts of a module corresponding to physical hardware.

There are seven **kinds** of circuit components.

> They are: nodes, wires, registers, output ports, input ports, submodule instances, and memories.

Author: If we don't distinguish between them, we end up having to make that distinction at another level somewhere else. It's better to treat them as distinct, even though they are tightly related to one another.

Radix-specified integer literals are only usable when constructing hardware
integer literals.  Any use in place of an integer is disallowed.

# Circuit Components

Circuit components are the named objects which may be contained within a module.

> They are: nodes, wires, registers, submodule instances, and ports.

Author: What else might be missing from this list?

Author: I was informed of `mem` earlier. Added a section on memories.

# Circuit Components

Circuit components are the named objects which may be contained within a module.
They are: nodes, wires, registers, ports, and submodule instances.

> TODO: What about `mem`\s?

Author: Have I captured all the circuit components here? 

I vaguely remember thinking there were 7 for some reason back in April. I know I counted `output` and `input` as separate, but that leaves me feeling like I'm missing one. Maybe I had included `mem`.

Author: It looks like `mem` is a circuit component. Added a section for it.

## Registers

Registers are the stateful elements of a design.

Registers can be connected (see [@sec:connects]).
The state of the register is controlled through what is connected to it.

The state of a register may be any passive type (see [@sec:passive-types]) and may not be `const`{.firrtl}.

> (TODO: is the `const` restriction a consequence of the passive part or not? Why is this required?)

Author: Is the `const` restriction a consequence of the passive part or not? Why is this required?

Intuitively, I get why it can't be `const`. Maybe it's just an additional restriction. But it might be important to motivate.

Author: @darthscsi explained that it's possible to get a time-varying `reg` in a circuit that only has `const` types.

Radix-specified integer literals are only usable when constructing hardware
integer literals.  Any use in place of an integer is disallowed.

# Circuit Components

Circuit components are the named parts of a module corresponding to physical hardware.

There are seven **kinds** of circuit components.
They are: nodes, wires, registers, output ports, input ports, submodule instances, and memories.

Circuit components can be connected together in the hardware (see [@sec:connects]).

Each circuit component in a module is associated
with a type called its type ([@sec:types]).
This is the type of data that travels across the component.
It is used to determine the legality of connections.

## Nodes

Nodes are named expressions in FIRRTL.

Example:

```firrtl
    node mynode = and(in, UInt<4>(1))

> ```

Reviewer 2: Nit: Can the PR use a consistent indenting of these examples? I tend to always left align them.

Ditto throughout.

The type of a node is the type of the expression given in the definition.

## Wires

Wires represent named expressions whose value is determined
by FIRRTL `connect`{.firrtl} statements (see [@sec:connects]).
Also unlike nodes, the type of a wire must be explicitly declared.

Example:

``` firrtl
    wire mywire: UInt<1>

    connect mywire, UInt<1>(0)
```

The type of a wire is given after the colon (`:`{.firrtl}).

## Registers

Registers are stateful elements of a design.

Registers can be connected (see [@sec:connects]).

> The state of the register is controlled through what is connected to it.

Reviewer 2: What is the "state" of a register? This is again on the intuition versus technical description issue.

The state of a register may be any non-`const`{.firrtl} passive type (see [@sec:passive-types]).
Registers are always associated with a clock.

> Optionally, registers may have a reset signal.

Reviewer 2: It may be simpler to state that FIRRTL has two types of registers: registers without a reset and registers with a reset.

On every cycle, a register will drive its current value

> and then latch the value it will take on for the next cycle.

Reviewer 2: "latch" is also possibly an inaccurate term that is overloaded with hardware latches.

The `regreset` keyword is used to declare a register with a reset.
The `reg` keyword is used to declare a register without a reset.

Examples:

``` firrtl
wire myclock : Clock
reg myreg : SInt, myclock
```

``` firrtl
wire myclock : Clock
reg myreg : SInt, myclock
```
```firrtl
wire myclock : Clock
wire myreset : UInt<1>
wire myinit : SInt
regreset myreg : SInt, myclock, myreset, myinit
```

For both variants of register, the type is given after the colon (`:`{.firrtl}).

Semantically, registers become flip-flops in the design.

> The the next value is latched on the positive edge of the clock.

Reviewer 2: ```suggestion
The next value is latched on the positive edge of the clock.
```

> For registers without a reset, the initial value is indeterminate (see [@sec:indeterminate-values]).

Reviewer 2: There was some very carefully constructed language in the original register statement section that intentionally does not talk about "initial value" and instead uses "reset value". This is extremely intentional as the initial value of a register is assumed to mean "the value of the register before any reset is applied" or "the value of the register at time zero".

Put differently, the initial value of both registers with and without resets is indeterminate.

Author: Whoops.

## Output Ports and Input Ports

The way a module interacts with the outside world is through its ports.

Example:

``` firrtl
    input myinput : UInt<1>
    output myinput : SInt<8>
```

> Ports are declared with the keywords `input` and `output`.

Reviewer 2: Nit throughout: there is a special syntax you can use to possible get syntax highlighting of fixed width text:

```
`input`{.firrtl}
```

The two keywords only differ in the rules for how they may be connected (see [@sec:connects]).

For both variants of port, the type is given after the colon (`:`{.firrtl}).

## Submodule Instances

A module in FIRRTL is allowed to contain submodules.
Each submodule instance must be given a name.
It must also name the FIRRTL module or external module it is instantiating.

Example:

```firrtl
inst passthrough of Passthrough
```

> This assumes you have a module named `Passthrough` declared elsewhere in your FIRRTL design.

Reviewer 2: ```suggestion
This assumes you have a module named `Passthrough` declared elsewhere in your FIRRTL `circuit`{.firrtl}.
```

It should be stated somewhere that modules may violate def-before-use.

Author: Added to my todo.

The keyword `of`{.firrtl} is used instead of the colon (`:`{.firrtl})
since `Passthrough` is not a type.

The type of a submodule instance is bundle type determined by its ports.
Each port creates a field in the bundle of the same name.
Among these fields, `output` ports are flipped, while `input` fields are unflipped.

For example:

```firrtl
module Passthrough :
  input in : UInt<8>
  output out : UInt<8>
  connect out, in
```

The type of the submodule instance `passthrough` above is thus:

```firrtl
{ flip in : UInt<8>, out : UInt<8> }
```

## Memories

> Memories are large stateful elements of a design.

Reviewer 2: Nit as with registers: memories may be small, too. They're yet-another stateful element.

## Passive Types

Stateful elements, such as registers and memories, may contain data of aggregate types.
Registers with bundle types are especially common.
However, when using bundle types in stateful elements, the notion of `flip` is meaningless.

> There is no directionality to the data inside a register; the data just *is*.

Author: Reworded to "does not make sense". My regards go out to Bruce.

## Passive Types

> Stateful elements, such as registers and memories, may contain data of aggregate types.

Reviewer: is "data" a precisely defined term?

Author: No.

The style that I'm going to start moving things towards is that the first sentence/paragraph of each section is informal motivation. A precise explanation of what "data" is or what it means to "contain data" would require we have rules regarding simulation of a FIRRTL design.

Registers with bundle types are especially common.
However, when using bundle types in stateful elements, the notion of `flip` is meaningless.

> There is no directionality to the data inside a register; the data just is.

Reviewer: ```suggestion
```

A passive type is a which does not make use of `flip`.
More precisely, a passive type is defined recursively:

* All ground types are passive.
* Probe types are passive.
* Properties types are passive.
* A vector type is passive if and only if the element type is passive.
* A bundle type is passive if and only if it contains no field which is marked `flip` and the type of each field is passive.
* Enums are passive if and only if each of their variants is passive. (TODO: Is this the proper terminology?)

Registers and memories may only be parametrized over passive types.

## Principal Types of Circuit Components

Each circuit component in a module is associated to a type called its principal type.
This is the type of data that flows through the component.

The principal type is used to determine whether two circuit components may be connected.

For wires, registers, ports, the principal component is declared after the colon `:` when the circuit component is declared.

For nodes, the principal type is inferred from the type of the expression on the left hand side of the equals sign `=` when the node is declared.

> For submodule instances, TODO: ???

Reviewer: is there a type other than Principal Type?

Author: @dtzSiFive asked me something similar. I added a sentence to clarify this.

tl;dr -- In something like `wire w : UInt<1>`, the *kind* is `wire` but the *principal type* is `UInt<1>`. It's just there to linguistically emphasize we are speaking formally.

Reviewer: Sorry why not then just say "type"?

## Passive Types

Stateful elements, such as registers and memories, may contain data of aggregate types.
Registers with bundle types are especially common.
However, when using bundle types in stateful elements, the notion of `flip` is meaningless.
There is no directionality to the data inside a register; the data just *is*.

A passive type is a which does not make use of `flip`.
More precisely, a passive type is defined recursively:

* All ground types are passive.
* All probe types are passive.

> * All properties types are passive.

Reviewer: Property types

# Statements

A module body consists of a sequence of statements.
Statements declare the circuit components and describe their connectivity.

## Connects

Circuit components can be connected together.

> This connection is both physical, as if the components were connected together with a wire,

Reviewer: ```suggestion
This connection can be both physical, as if the components were connected together with a wire,
```

Author: Thanks.

## Passive Types

Stateful elements, such as registers and memories, may contain data of aggregate types.
Registers with bundle types are especially common.
However, when using bundle types in stateful elements, the notion of `flip` is meaningless.
There is no directionality to the data inside a register; the data just *is*.

> A passive type is a which does not make use of `flip`.

Reviewer 2: ```suggestion
A passive type is a type which does not make use of `flip`.
```

More precisely, a passive type is defined recursively:

* All ground types are passive.
* All probe types are passive.
* All property types are passive.
* A vector type is passive if and only if the element type is passive.
* A bundle type is passive if and only if it contains no field which is marked `flip`
  and the type of each field is passive.

> * An enum type is passive if and only if the type of each of its variants is passive.

Reviewer 2: Mega-nit: I think the rest of the spec uses `-` for itemized lists. This doesn't affect the output, just causes my OCD to kick in.

Author: Changed.

> Registers and memories may only be parametrized over passive types.

Reviewer 2: Can the PR avoid use of "parameterized". This will eventually mean something else.

# Statements

A module body consists of a sequence of statements.
Statements declare the circuit components and describe their connectivity.

## Connects

The components of a module can be connected together using connect statements.

These connections can be both physical, as if the components were connected together with a wire,
but also logical, allowing metadata such as probes (see [@sec:probe-types;@sec:probe])

> and properties (see [@sec:property-types]) to flow across it during elaboration.

Reviewer 2: I'd avoid "physical" and "logical" as that introduces new, undefined concepts. Is there a way to do this by talking about different connections for hardware and non-hardware components?

Author: There's no formal distinction, so let's just fudge it until there is.

### Registers without Reset

> TODO: Do I remove this?

Reviewer: why would we remove this?

Author: In the next few weeks, I want to reorganize a bunch of things in the doc.

The chapter on "Statements" is not the proper grouping. It maybe made sense at one point. But this section is doing much more work than it should be. 

Rather than going statement-by-statement, I am going to move us towards explaining things concept-by-concept. The section I added on Circuit Components causes some tension, begging material from the Statements section to be moved up front. 

Until I have the re-organization finished, some material might get split up (with appropriate cross-references).

during transformational passes. See [@sec:conditionals] for details on the
conditional statement.

## Wire

See [@sec:wires].

> ## Registers

Reviewer: If you are changing these titles to be named after the keyword, which I don't think really makes sense, shouldn't this one be called Reg?

## Invalidates

> An invalidate statement is used to indicate that a circuit component contains indeterminate values (see [@sec:indeterminate-values]).

Reviewer: it seems this is purely a whitespace change? This doc has been formatted for 80 character limit, is there a reason you are changing this line?

Author: It looks like I removed the second sentence. (And this was outdated by another change, so just keeping that in mind).

I would like to gradually shift over to using a git-friendly formatting, with one(-ish) sentence per line.

Author: I'll also ensure I have a reasonable line length. I just realized my editor settings don't make it obvious for me to tell when I hit the 80 or 100 col line limit.

Author: I pushed a change to cap all of my changes at 100 cols.

Hopefully that isn't controversial.

invalidate w
```

Invalidate statements can be applied to any circuit component of any

> type. However, if the circuit component cannot be connected to, then the

Author: As far as the rest of the spec indicates, no circuit component cannot be connected to.

Arguably submodule instances or memories were initially considered, but it is not explicitly forbidden. (Or if it is, we need to reference where and why).

Reviewer: >  no circuit component cannot be connected to.

That is not true, you are allowed to not connect to e.g. the `input` of your top level module in the circuit.

Author: Ah. Thank you. I was trying to work out when this would be restricted.

Is there any problem with me changing this to a component that has Sink or Duplex flow? "Cannot be connected to" puts too much obligation on the reader to read between the lines.

### Nested Declarations

If a circuit component is declared within a conditional statement,

> connections to the it are unaffected by the condition.

Reviewer: "to the it" --> something wrong

Author: Thanks.

connect w.b, x.b
```

> ## Mem

Reviewer: I don't think we should change this heading...?

# Expressions

FIRRTL expressions are used for creating constant integers,
for creating literal property type expressions,
for referencing a circuit component,
for statically and dynamically accessing a nested element within a component,
for creating multiplexers,

> for performing primitive operations, and

Reviewer 2: Mega-nit: unnecessary manual text flowing.

# Width Inference

For all circuit components declared with unspecified widths,
the FIRRTL compiler will infer the minimum possible width that
maintains the legality of all its incoming connections.
If a circuit component has no incoming connections, and the width is unspecified,

> then an error is thrown to indicate that the width could not be inferred.

Reviewer 2: Nit: unnecessary text reflowing.

       | "o" | "p" | "q" | "r" | "s" | "t" | "u"
       | "v" | "w" | "x" | "y" | "z" ;

(* Fileinfo communicates Chisel source file and line/column info *)
linecol = digit_dec , { digit_dec } , ":" , digit_dec , { digit_dec } ;

> lineinfo = string, " ", linecol ;

Author: Good catch. The merge messed this up.
